File: frontend/src/logging.ts
// Frontend log capture: hooks console.* and forwards entries to the backend
// log broker over the existing WebSocket. Works standalone in the browser too
// (no Tauri dependency). Keeps a local ring so the GUI log panel shows
// frontend-originated entries even before the socket connects.

export interface LogEntry {
  level: "debug" | "info" | "warn" | "error";
  message: string;
  source: "frontend" | "shell" | "backend";
  ts: number;
}

type Send = (env: { type: string; payload: Record<string, unknown> }) => void;

const LOCAL_RING: LogEntry[] = [];
const MAX_RING = 500;

let sendFn: Send | null = null;
let installed = false;

function push(level: LogEntry["level"], message: string) {
  const entry: LogEntry = { level, message, source: "frontend", ts: Date.now() / 1000 };
  LOCAL_RING.push(entry);
  if (LOCAL_RING.length > MAX_RING) LOCAL_RING.shift();
  try {
    sendFn?.({ type: "log", payload: { level, message, source: "frontend" } });
  } catch {
    /* ignore */
  }
}

/** Install console.* forwarding. Call once, after the socket hook is alive. */
export function installFrontendLogging(send: Send): () => void {
  sendFn = send;
  if (installed) return () => {};
  installed = true;

  const orig = {
    log: console.log,
    info: console.info,
    warn: console.warn,
    error: console.error,
    debug: console.debug,
  };

  const fmt = (...args: unknown[]): string =>
    args.map((a) => (typeof a === "string" ? a : safeStringify(a))).join(" ");

  console.log = (...a) => { orig.log(...a); push("info", fmt(...a)); };
  console.info = (...a) => { orig.info(...a); push("info", fmt(...a)); };
  console.warn = (...a) => { orig.warn(...a); push("warn", fmt(...a)); };
  console.error = (...a) => { orig.error(...a); push("error", fmt(...a)); };
  console.debug = (...a) => { orig.debug(...a); push("debug", fmt(...a)); };

  window.addEventListener("error", (e) => push("error", `window error: ${e.message}`));
  window.addEventListener("unhandledrejection", (e) => push("error", `unhandled rejection: ${String(e.reason)}`));

  return () => {
    installed = false;
    console.log = orig.log;
    console.info = orig.info;
    console.warn = orig.warn;
    console.error = orig.error;
    console.debug = orig.debug;
    sendFn = null;
  };
}

export function getLocalLogs(limit = 200): LogEntry[] {
  return LOCAL_RING.slice(-limit);
}

function safeStringify(value: unknown): string {
  try {
    return JSON.stringify(value);
  } catch {
    return String(value);
  }
}

File: src-tauri/src/logging.rs
//! Shell-side logging: console + rotating file + WS forwarding + GUI ring.
//!
//! Every `log` record produced by the Rust shell is written to
//! `~/.jarvis/logs/jarvis-shell.log` (with size-based rollover), kept in an
//! in-memory ring for the GUI (`get_logs` command), and forwarded to the
//! backend over WebSocket as a `log` envelope when connected. The backend
//! aggregates shell/backend/frontend logs in one broker.

use std::collections::VecDeque;
use std::fs::{File, OpenOptions};
use std::io::Write;
use std::path::PathBuf;
use std::sync::{Mutex, OnceLock};

use log::{LevelFilter, Log, Metadata, Record};
use serde::Serialize;

const MAX_FILE_BYTES: u64 = 5 * 1024 * 1024;

/// One log entry exposed to the GUI and forwarded over WS.
#[derive(Debug, Clone, Serialize)]
pub struct LogEntry {
    pub level: String,
    pub message: String,
    pub source: String,
    pub ts: f64,
}

struct JarvisLogger {
    file: Mutex<Option<File>>,
    ring: Mutex<VecDeque<LogEntry>>,
    max_ring: usize,
    app: OnceLock<tauri::AppHandle>,
}

static LOGGER: OnceLock<JarvisLogger> = OnceLock::new();

fn logger() -> &'static JarvisLogger {
    LOGGER.get_or_init(|| JarvisLogger {
        file: Mutex::new(None),
        ring: Mutex::new(VecDeque::new()),
        max_ring: 500,
        app: OnceLock::new(),
    })
}

fn log_dir() -> PathBuf {
    let home = std::env::var_os("HOME").map(PathBuf::from).unwrap_or_else(std::env::temp_dir);
    home.join(".jarvis").join("logs")
}

fn open_file() -> Option<File> {
    let dir = log_dir();
    std::fs::create_dir_all(&dir).ok()?;
    let path = dir.join("jarvis-shell.log");
    // Roll over if the current file is too big.
    if std::fs::metadata(&path).map(|m| m.len() > MAX_FILE_BYTES).unwrap_or(false) {
        let _ = std::fs::rename(&path, dir.join("jarvis-shell.log.1"));
    }
    OpenOptions::new().create(true).append(true).open(path).ok()
}

fn now_epoch_ms() -> f64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or(0.0)
}

/// Best-effort masking of obvious key prefixes. The backend redacts again, so
/// this is purely defense in depth.
fn redact(msg: &str) -> String {
    let mut out = msg.to_string();
    for marker in ["sk-", "csk-", "ghp_"] {
        let mut start = 0;
        loop {
            let Some(rel) = out[start..].find(marker) else { break };
            let abs = start + rel;
            let rest = &out[abs + marker.len()..];
            let end = rest.find(|c: char| !c.is_ascii_alphanumeric()).unwrap_or(rest.len());
            if end >= 12 {
                out.replace_range(abs + marker.len()..abs + marker.len() + end, "***");
            }
            start = abs + marker.len();
        }
    }
    out
}

fn format_line(record: &Record) -> String {
    let level = record.level().to_string();
    let msg = redact(&record.args().to_string());
    format!("{:.3} {level:<5} {} {}", now_epoch_ms(), record.target(), msg)
}

impl Log for JarvisLogger {
    fn enabled(&self, metadata: &Metadata) -> bool {
        metadata.level() <= log::max_level()
    }

    fn log(&self, record: &Record) {
        if !self.enabled(record.metadata()) {
            return;
        }
        let line = format_line(record);

        if let Ok(mut guard) = self.file.lock() {
            if guard.is_none() {
                *guard = open_file();
            }
            if let Some(file) = guard.as_mut() {
                let _ = writeln!(file, "{line}");
            }
        }

        if let Ok(mut ring) = self.ring.lock() {
            ring.push_back(LogEntry {
                level: record.level().to_string().to_lowercase(),
                message: line.clone(),
                source: "shell".into(),
                ts: now_epoch_ms() / 1000.0,
            });
            while ring.len() > self.max_ring {
                ring.pop_front();
            }
        }

        // Forward to the backend broker over WS if connected.
        if let Some(app) = self.app.get() {
            let payload = serde_json::json!({
                "level": record.level().to_string().to_lowercase(),
                "message": line,
                "source": "shell",
            });
            crate::ws::send(app, "log", &payload);
        }
    }

    fn flush(&self) {
        if let Ok(guard) = self.file.lock() {
            if let Some(file) = guard.as_ref() {
                let _ = file.sync_all();
            }
        }
    }
}

/// Initialize the global shell logger. Call once at startup.
pub fn init() {
    let filter = std::env::var("RUST_LOG")
        .ok()
        .and_then(|v| v.parse::<LevelFilter>().ok())
        .unwrap_or(LevelFilter::Info);
    log::set_logger(logger()).expect("logger already set");
    log::set_max_level(filter);

    if let Ok(mut guard) = logger().file.lock() {
        *guard = open_file();
    }
    log::info!("JARVIS shell logger initialized (level={filter})");
}

/// Give the logger a handle so records can be forwarded to the backend.
/// Called from `.setup()` once the app handle exists.
pub fn attach(app: &tauri::AppHandle) {
    let _ = logger().app.set(app.clone());
    log::debug!("shell logger attached to app handle");
}

/// Tauri command: return recent shell log entries for the GUI log panel.
#[tauri::command]
pub fn get_logs(limit: Option<usize>) -> Vec<LogEntry> {
    let n = limit.unwrap_or(200);
    logger().ring.lock().map(|r| r.iter().rev().take(n).cloned().collect()).unwrap_or_default()
}

/// Tauri command: wipe the in-memory ring.
#[tauri::command]
pub fn clear_logs() {
    if let Ok(mut ring) = logger().ring.lock() {
        ring.clear();
    }
}
